### PR TITLE
Make 'Rect' a 'constexpr' class

### PR DESCRIPTION
**Pull request by vittorioromeo**
## Description

This PR turns `sf::Rect` into a `constexpr`-friendly class, and also provides a few more improvements:
- Removes dependency on `<algorithm>` from a widely used header, improving compilation time
    - Some other headers were relying on transitive includes from `Rect.hpp`, so they have been fixed
- Removes unnecessary `sf::` qualification when referring to `Vector` inside `Rect` 

## Tasks

* [ ] Tested on Linux
* [x] Tested on Windows
* [ ] Tested on macOS
* [ ] Tested on iOS
* [ ] Tested on Android

## How to test this PR?

CI.

### CODE REVIEW

**Comment by Bromeon** [comment]
Maybe before commenting on `constexpr` for `Rect` and `Vertex` in particular:

We have guy who put in all the effort at the beginning of the year in https://github.com/SFML/SFML/pull/1741 and even split the commits by class. I know his PR might not be ready-to-use now, but I'm not too happy about telling him we still need to discuss the topic, and then later rewriting everything without him, with no single credit. This could make him reconsider future contributions, and rightfully so.

**Comment by vittorioromeo** [comment]
> Maybe before commenting on `constexpr` for `Rect` and `Vertex` in particular:
> 
> We have guy who put in all the effort at the beginning of the year in #1741 and even split the commits by class. I know his PR might not be ready-to-use now, but I'm not too happy about telling him we still need to discuss the topic, and then later rewriting everything without him, with no single credit. This could make him reconsider future contributions, and rightfully so.
> 
> What we could do is take his commits (under his author), perform the necessary changes and add `Signed-off-by:` clause in the commit message.

Thanks for bringing this up, I do believe it is an important thing to consider.

I'm totally happy for @fgallegosalido to take the credit for these changes, which are pretty much the same as the ones he contributed back in #1741. I am not sure what the easiest way to achieve that would be, but I also don't really feel like dealing with `git` cherry-picking and rebasing before we can merge these changes in.

Is it possible to force-change the author of a commit? I'd be fine with changing the authorship of my commit to @fgallegosalido. Alternatively, if he is fine with it, I'd also propose simply adding *"Originally contributed to @fgallegosalido"* to these commits or something alongside those lines to avoid any `git`-related pain/maintenance.

Let me know what you think.

**Comment by Bromeon** [comment]
An alternative would be a `Co-authored-by:` clause in the commit message. The semantics are a bit different: this means that multiple developers authored a change, while `Signed-off-by:` means that one person authored it and another "signed it off", i.e. adjusted it to latest `master`, made minor corrections etc. "Sign off" could also have the literal meaning of just approving it.

Maybe @fgallegosalido sees this himself and can tell us what he prefers? Otherwise I'd say we go with one of those two clauses (they are "official" in the sense that GitHub and other tools can parse them).

**Comment by Bromeon** [comment]
Thank you @fgallegosalido!

@vittorioromeo: In that case, it's probably easiest if you leave this PR as-is and simply change the commit message to include:
```
Co-authored-by: fgallegosalido <fgallegosalido@gmail.com>
```

I'll try to review the technical parts later today 🙂

**Comment by vittorioromeo** [comment]
Added! :)

**Comment by fgallegosalido** [comment]
I also think we can close #1741, as these wave of changes are much more complete than what I did back then. Thank you @vittorioromeo for your contributions! We all wanted SFML to be modernized ^.^

File: include/SFML/Graphics/Rect.inl
Comment (vittorioromeo) [reply include/SFML/Graphics/Rect.inl:88]: It is not a requirement here. Making local variables `const` is beneficial in my opinion, it helps whoever is reading the source code to understand what the moving parts of a function are. For by-value parameters, the declaration can omit the `const` and the definition can have it for the same function -- that is valid C++.

I don't think we lose anything with this small change, and we promote the use of `const` which does help increase readability of function bodies.

File: include/SFML/Graphics/Rect.inl
Comment (vittorioromeo) [reply include/SFML/Graphics/Rect.inl:117]: This is my deliberate style choice for writing lambdas. We don't have a style guide yet, so I don't really think it would be beneficial to find consensus on style during this PR. As a reminder, our priority regarding style should be to finalize #1861 rather than to perform small changes on a per-PR basis. 

Once we have `clang-format` in place, many of the decisions made here might not be relevant anymore. I don't think it's worthwhile use of our time to discuss minutiae like a single whitespace here, especially since we don't have existing practice to use as reference.

Comment (Bromeon) [reply include/SFML/Graphics/Rect.inl:117]: Thought it might be a typo -- but fully agree, let's push those things to the formatting PR. 👍

File: include/SFML/Graphics/Rect.inl
Comment (Bromeon) [reply include/SFML/Graphics/Rect.inl:88]: Yeah, it's probably another case of "we might want to decide how to do this SFML-wide" or at least decide to leave it up to the code location 🙂

So let's keep it for now.

**Comment by vittorioromeo** [comment]
@Bromeon: ping :)

**Review by Bromeon — APPROVED** [review]

File: include/SFML/Graphics/Rect.inl
Comment (ChrisThrasher) [inline include/SFML/Graphics/Rect.inl:80]: image

lol (not upset, just having a laugh)

**Pull request merged**
